### PR TITLE
Fix building on OSX

### PR DESCRIPTION
**Pull request by enoposix**
This doesn't fix linking as I'm not sure how to do that with cmake.

```
Scanning dependencies of target arc_unpacker
[ 57%] Building CXX object CMakeFiles/arc_unpacker.dir/src/main.cc.o
[ 57%] Linking CXX executable arc_unpacker
clang: warning: argument unused during compilation: '-pthread'
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      convert_locale(au::bstr const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allo
cator<char> > const&) in locale.cc.o
  "_iconv_close", referenced from:
      convert_locale(au::bstr const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allo
cator<char> > const&) in locale.cc.o
  "_iconv_open", referenced from:
      convert_locale(au::bstr const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allo
cator<char> > const&) in locale.cc.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [arc_unpacker] Error 1
make[1]: *** [CMakeFiles/arc_unpacker.dir/all] Error 2
make: *** [all] Error 2
```

I haven't tested it on other systems (linux or windows).

### CODE REVIEW

File: src/algo/call_stack_keeper.cc [hunk 1]
```diff
@@ -29,8 +30,9 @@ bool CallStackKeeper::recursion_limit_reached() const
 
 void CallStackKeeper::recurse(const std::function<void()> &action)
 {
+    const std::string recursion_error = std::string("Recursion limit reached");
```
Comment (rr-) [inline src/algo/call_stack_keeper.cc]: Consider using `const auto recursion_error = std::string(...)` or `const std::string recursion_error = ...`. (Do we have `"..."s` syntax in C++14?)

File: src/algo/call_stack_keeper.cc [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #include "algo/call_stack_keeper.h"
 #include "err.h"
+#include <string>
```
Comment (rr-) [inline src/algo/call_stack_keeper.cc]: Please run `./tools/checkstyle` and fix include order (`<>` before `""`, includes are sorted lexicographically, relevant header is included before all else).

File: src/dec/almond_collective/pac3_archive_decoder.cc [hunk 3]
```diff
@@ -3,6 +3,7 @@
 #include "algo/locale.h"
 #include "algo/range.h"
 #include "io/memory_byte_stream.h"
+#include <array>
```
Comment (rr-) [inline src/dec/almond_collective/pac3_archive_decoder.cc]: Include order

File: src/dec/dogenzaka/a_file_decoder.cc [hunk 4]
```diff
@@ -2,6 +2,7 @@
 #include <algorithm>
 #include "algo/crypt/sha1.h"
 #include "algo/range.h"
+#include <array>
```
Comment (rr-) [inline src/dec/dogenzaka/a_file_decoder.cc]: Include order

File: src/dec/entis/eri_image_decoder.cc [hunk 5]
```diff
@@ -41,8 +42,8 @@ static image::EriHeader read_header(
 
     const s32 width         = input_stream.read_le<u32>();
     const s32 height        = input_stream.read_le<u32>();
-    header.width            = std::abs(width);
-    header.height           = std::abs(height);
+    header.width            = std::fabs(width);
+    header.height           = std::fabs(height);
```
Comment (rr-) [inline src/dec/entis/eri_image_decoder.cc]: `s32` are signed 32-bit integers, why are we using `float` abs here?

File: src/dec/fc01/mcg_image_decoder.cc [hunk 6]
```diff
@@ -35,9 +36,9 @@ static bstr transform_v200(
         int p00 = planes[i][pos];
         int p10 = planes[i][pos + width] - p00;
         int p01 = planes[i][pos + 1] - p00;
-        p00 = std::abs(p01 + p10);
-        p01 = std::abs(p01);
-        p10 = std::abs(p10);
+        p00 = std::fabs(p01 + p10);
+        p01 = std::fabs(p01);
+        p10 = std::fabs(p10);
```
Comment (rr-) [inline src/dec/fc01/mcg_image_decoder.cc]: See the other comment about `fabs`

File: src/dec/microsoft/bmp_image_decoder.cc [hunk 7]
```diff
@@ -74,7 +75,7 @@ static Header read_header(io::BaseByteStream &input_stream)
         h.palette_size = header_stream.read_le<u32>();
         h.important_colors = header_stream.read_le<u32>();
     }
-    h.height = std::abs(height);
+    h.height = std::fabs(height);
```
Comment (rr-) [inline src/dec/microsoft/bmp_image_decoder.cc]: See the other comment about `fabs`

File: src/dec/entis/eri_image_decoder.cc (hunk 5 above)
Comment (joshdekock) [inline src/dec/entis/eri_image_decoder.cc]: `std::abs` results in an ambigious overload apparently, so the workaround is to use `std::fabs` or std::abs with a cast to double beforehand.

Comment (rr-) [inline src/dec/entis/eri_image_decoder.cc]: It should use these headers: http://en.cppreference.com/w/cpp/numeric/math/abs

Comment (kxz) [inline src/dec/entis/eri_image_decoder.cc]: `#include <cstdlib>` fixes this without having to change `std::abs` to `std::fabs`.

**Comment by kxz** [comment]
The linking issue is fixed by passing `-DICONV_LIBRARY=iconv` to `cmake`.

**Comment by rr-** [comment]
Added support upstream loosely based on the PR. Additionally, added new Travis target. Thanks for all users involved.

**Comment by enoposix** [comment]
Can confirm that this fixes compilation entirely on OSX.

**Pull request closed without merging**